objects/src/accounts/seed.rs

```diff
@@ -0,0 +1,244 @@
+use super::{compute_digest, AccountError, AccountId, AccountType, Digest, Felt, Vec, Word};
+
+#[cfg(feature = "concurrent")]
+use std::{
+    sync::{
+        mpsc::{self, Sender},
+        OnceLock,
+    },
+    thread::{self, spawn},
+};
+
+#[cfg(feature = "log")]
+mod log {
+    use super::{
+        super::{digest_pow, Digest, Word},
+        AccountId, AccountType,
+    };
+    use assembly::utils::to_hex;
+    use crypto::utils::string::String;
+    use crypto::FieldElement;
+
+    /// Keeps track of the best digest found so far and count how many iterations have been done.
+    pub struct Log {
+        digest: Digest,
+        seed: Word,
+        count: usize,
+        pow: u32,
+    }
+
+    /// Given a [Digest] returns its hex representaiton.
+    pub fn digest_hex(digest: Digest) -> String {
+        to_hex(&digest.as_bytes()).expect("hex formatting failed")
+    }
+
+    /// Given a [Word] returns its hex representation.
+    pub fn word_hex(word: Word) -> String {
+        to_hex(FieldElement::elements_as_bytes(&word)).expect("hex formatting failed")
+    }
+
+    impl Log {
+        pub fn start(
+            digest: Digest,
+            seed: Word,
+            account_type: AccountType,
+            on_chain: bool,
+        ) -> Self {
+            log::info!(
+                "Generating new account seed [pow={}, digest={}, seed={} type={:?} onchain={}]",
+                digest_pow(digest),
+                digest_hex(digest),
+                word_hex(seed),
+                account_type,
+                on_chain,
+            );
+
+            Self {
+                digest,
+                seed,
+                count: 0,
+                pow: 0,
+            }
+        }
+
+        pub fn iteration(&mut self, digest: Digest, seed: Word) {
+            self.count += 1;
+
+            let pow = digest_pow(digest);
+            if pow >= self.pow {
+                self.digest = digest;
+                self.seed = seed;
+                self.pow = pow;
+            }
+
+            if self.count % 500_000 == 0 {
+                log::debug!(
+                    "Account seed loop [count={}, pow={}, digest={}, seed={}]",
+                    self.count,
+                    self.pow,
+                    digest_hex(self.digest),
+                    word_hex(self.seed),
+                );
+            }
+        }
+
+        pub fn done(self, digest: Digest, seed: Word, account_id: AccountId) {
+            log::info!(
+                "Found account seed [pow={}, current_digest={}, current_seed={} type={:?} onchain={}]]",
+                digest_pow(digest),
+                digest_hex(digest),
+                word_hex(seed),
+                account_id.account_type(),
+                account_id.is_on_chain(),
+            );
+        }
+    }
+}
```

Comment:
I would probably have the logging module at the bottom of the file.

Reply:
moved